### PR TITLE
[WIP] DSL tutorial

### PR DESCRIPTION
**Pull request by bollu**

### CODE REVIEW

File: lean/main/dsls.lean [hunk 1]
```diff
@@ -0,0 +1,514 @@
...
+we would write out an "interpretation" macro such as `[imp_aexp|...]`,
+and the declare the translation as a macro rule:
+-/
+
+syntax "[imp_aexp|" imp_aexp "]" : term
+macro_rules
+| `([imp_aexp| $n:num ]) => do
+   let n : Nat := n.toNat -- grab the number.
+   let nStx : Syntax := Lean.quote n -- quote the number.
+   `(AExp.ANat $(nStx)) -- return the syntax.
```
Comment (arthurpaulino) [inline lean/main/dsls.lean:118]: This should work, right?
```suggestion
| `([imp_aexp| $n:num ]) => `(AExp.ANat n)
```

File: lean/main/dsls.lean [hunk 2]
```diff
@@ -0,0 +1,514 @@
...
+We go directly to parsing with `elab`, since the `macro_rules`
+approach does not have anything more interesting to say.
+-/
+
+def elab_aexp_plus (s: Syntax): TermElabM Expr := 
+match s with 
+| `(imp_aexp| $x:imp_aexp + $y:imp_aexp) => do 
+     -- recrsively expand xExpr, yExpr via `Term.elabTerm`
+     let xExpr <- Term.elabTerm (<- `([imp_aexp'| $x])) (expectedType? := none)
+     let yExpr <- Term.elabTerm (<- `([imp_aexp'| $y])) (expectedType? := none)
```
Comment (arthurpaulino) [inline lean/main/dsls.lean:270]: Can we use `elab_AExp_num` here? I don't want to be reliant on `Term.elabTerm`

Notice that I didn't use `Term.elabTerm` to build the DSL [here](https://github.com/arthurpaulino/FxyLang/blob/master/MetaDebug.lean)

File: lean/main/dsls.lean [hunk 3]
```diff
@@ -0,0 +1,514 @@
...
+     | n => mkAppM ``BExp.BVar #[mkStrLit str]
+
+/-
+To elaborate the less than (`<`) operator on `aexp`s, we wrie a helper called
+`elab_aexp`, that calls `elabTerm` on the term `[imp_aexp'| $s]`. This produces
+an `Expr` node, which we use to build a `BExp.BLess`
+-/
+| `(imp_bexp| $x:imp_aexp < $y:imp_aexp) => do  
+     let elab_aexp (s: Syntax): TermElabM Expr := do
+        Term.elabTerm (<- `([imp_aexp'| $s])) none
```
Comment (arthurpaulino) [inline lean/main/dsls.lean:341]: Same issue as before. Can we do it without `Term.elabTerm`?

**Review by arthurpaulino — CHANGES_REQUESTED** [review]

File: lean/main/dsls.lean (hunk 1 above)
Comment (bollu) [reply lean/main/dsls.lean:118]: Ah yes, that makes much more sense, than to pull out data and then quote again :D

File: lean/main/dsls.lean (hunk 2 above)
Comment (bollu) [reply lean/main/dsls.lean:270]: I don't believe it's possible. calling `elab_AExp_num` will mean that when elaborating something like `1 + (2 + 3)`, we try to elaborate `yExpr = (2 + 3)` using `elab_AExp_num` which will not work.

Essentially, we use `Term.elabTerm` to give us the ability to recursively expand the _syntax category_ `imp_aexp`, whose elaboration is being defined piece-by-piece.

File: lean/main/dsls.lean (hunk 3 above)
Comment (bollu) [reply lean/main/dsls.lean:338]: Same answer as before, I don't think it works without `Term.elabTerm`.

**Pull request merged**